### PR TITLE
Add support for Razer Basilisk V2

### PR DESCRIPTION
**Pull request by mahmoudhossam**
This is basically a copy of #1142 with all conflicts resolved and @z3ntu's comments addressed.

### CODE REVIEW

**Comment by mahmoudhossam** [comment]
There seems to be a problem with an attribute: `tilt_hwheel`

As I'm new to the project, I'm open for suggestions on how to fix that.

**Comment by nightsky30** [comment]
> There seems to be a problem with an attribute: `tilt_hwheel`
> 
> As I'm new to the project, I'm open for suggestions on how to fix that.

Looks like the fake driver `razerbasiliskv2.cfg` needs the same attribute values as were included in #1321. See:

https://github.com/openrazer/openrazer/blob/6c415bb6843cab690538c41755eb004d6db624f1/pylib/openrazer/_fake_driver/razerbasiliskultimatewired.cfg#L40-L44

Also, #1321 included the associated changes in the script that handles the tilt wheel attribute values when generating fake drivers:

https://github.com/openrazer/openrazer/blob/6c415bb6843cab690538c41755eb004d6db624f1/scripts/generate_fake_driver.sh#L72-L76

Not sure if you need to modify the script, or if simply adding the values to the .cfg file will suffice for the CI checks. You could give either/or both a shot :)

**Comment by mahmoudhossam** [comment]
@nightsky30 Thank you very much!

The tests now pass and this is ready to be reviewed/merged.

**Comment by jgrim** [comment]
Could I build this by running the make file on this branch and be able to install it before it's public release? Is there anything special I should know before I attempt that?

**Comment by mahmoudhossam** [comment]
@jgrim you can test it yourself, I haven't received any reviews on this PR yet so I can't really tell as this is my first interaction with the code.

**Comment by kerberizer** [comment]
@jgrim I've been using @mahmoudhossam 's branch without problems for a few weeks now, with two caveats however:
* I'm on Arch, so I used [openrazer-git](https://aur.archlinux.org/pkgbase/openrazer-git/) from AUR (i.e. haven't build it “manually”);
* I used it only to set the LED colors for the in-memory profiles—which works—but not for much else.

**Comment by jgrim** [comment]
@kerberizer @mahmoudhossam Thanks a lot. I'll hold off then until it's reviewed. :bow:

File: driver/razermouse_driver.c [hunk 1]
```diff
@@ -3359,48 +3446,339 @@ static DEVICE_ATTR(right_matrix_effect_none,        0220, NULL,
 static DEVICE_ATTR(backlight_led_state,            0660, razer_attr_read_backlight_led_state, razer_attr_write_backlight_led_state);
 
 
+#define REP4_DPI_UP  0x20
+#define REP4_DPI_DN  0x21
+#define REP4_TILT_L  0x22
+#define REP4_TILT_R  0x23
+#define REP4_PROFILE 0x50
+#define REP4_SNIPER  0x51
```
Comment (z3ntu) [inline driver/razermouse_driver.c:3571]: Where do these names come from?

File: daemon/openrazer_daemon/hardware/mouse.py [hunk 2]
```diff
@@ -2006,6 +2006,64 @@ def _resume_device(self):
...
+               'set_scroll_static_naga_hex_v2', 'set_scroll_spectrum_naga_hex_v2', 'set_scroll_none_naga_hex_v2', 'set_scroll_reactive_naga_hex_v2', 'set_scroll_breath_random_naga_hex_v2', 'set_scroll_breath_single_naga_hex_v2', 'set_scroll_breath_dual_naga_hex_v2',
+               # Can set LOGO and Scroll with custom
+               'set_custom_effect', 'set_key_row']
+
+    DEVICE_IMAGE = "https://assets.razerzone.com/eeimages/support/products/1617/1617_basilisk-v2.png"
+
+    # Deprecated - RAZER_URLS be removed in future.
+    RAZER_URLS = {
+        "top_img": DEVICE_IMAGE
+    }
```
Comment (z3ntu) [inline daemon/openrazer_daemon/hardware/mouse.py]: This block can be removed

File: README.md [hunk 3]
```diff
@@ -121,6 +121,7 @@ The devices below are fully feature supported by OpenRazer, which means all avai
 | Razer Basilisk X HyperSpeed                   |  1532:0083  |
 | Razer DeathAdder V2                           |  1532:0084  |
 | Razer Viper Mini                              |  1532:008A  |
+| Razer Basilisk V2                             |  1532:0085  |
```
Comment (z3ntu) [inline README.md:128]: Move the line one up, so it's sorted: `0084`, `0085`, `008A`

File: driver/razerchromacommon.c [hunk 4]
```diff
@@ -1068,8 +1068,8 @@ struct razer_report razer_chroma_misc_set_dpi_xy(unsigned char variable_storage,
     struct razer_report report = get_razer_report(0x04, 0x05, 0x07);
 
     // Keep the DPI within bounds
-    dpi_x = clamp_u16(dpi_x, 128, 16000);
-    dpi_y = clamp_u16(dpi_y, 128, 16000);
+    dpi_x = clamp_u16(dpi_x, 100, 20000);
+    dpi_y = clamp_u16(dpi_y, 100, 20000);
```
Comment (z3ntu) [inline driver/razerchromacommon.c]: Please rebase on master, this part has been merged already

File: driver/razermouse_driver.h [hunk 5]
```diff
@@ -62,6 +62,7 @@
 #define USB_DEVICE_ID_RAZER_BASILISK_X_HYPERSPEED 0x0083
 #define USB_DEVICE_ID_RAZER_DEATHADDER_V2 0x0084
 #define USB_DEVICE_ID_RAZER_VIPER_MINI 0x008A
+#define USB_DEVICE_ID_RAZER_BASILISK_V2 0x0085
```
Comment (z3ntu) [inline driver/razermouse_driver.h:66]: move this up so it's sorted as well

File: install_files/udev/99-razer.rules [hunk 6]
```diff
@@ -5,7 +5,7 @@ GOTO="razer_end"
 LABEL="razer_vendor"
 
 # Mice
-ATTRS{idProduct}=="0013|0016|0020|0024|0025|002e|002f|0032|0034|0036|0037|0038|0039|0040|0041|0042|0043|0044|0045|0046|0048|004c|004f|0050|0053|0054|0059|005a|005b|005c|005e|0060|0062|0064|0067|006a|006b|006c|006e|006f|0070|0071|0072|0073|0078|007a|007b|0083|0084|008a", \
+ATTRS{idProduct}=="0013|0016|0020|0024|0025|002e|002f|0032|0034|0036|0037|0038|0039|0040|0041|0042|0043|0044|0045|0046|0048|004c|004f|0050|0053|0054|0059|005a|005b|005c|005e|0060|0062|0064|0067|006a|006b|006c|006e|006f|0070|0071|0072|0073|0078|007a|007b|0083|0084|008a|0085", \
```
Comment (z3ntu) [inline install_files/udev/99-razer.rules]: move the new entry to the left one item so it's sorted

**Review by z3ntu — CHANGES_REQUESTED** [review]
A few comments. I am not a fan of this PR because it has a bunch of custom new code for the tilt wheel but it seems it's needed... I also can't verify most of the behavior here as I don't have the mouse

File: driver/razermouse_driver.c (hunk 1 above)
Comment (mahmoudhossam) [reply driver/razermouse_driver.c:3571]: If the names are any indicator: those are for tilting the wheel left and right, the profile switch button, and the detachable "sniper" button.

**Comment by jgrim** [comment]
I built and installed this and the left and right scrolling doesn't appear to work. I was able to set colors and such using RazerGenie but that's about it.

**Comment by jgrim** [comment]
> I built and installed this and the left and right scrolling doesn't appear to work. I was able to set colors and such using RazerGenie but that's about it.

I got it to work by setting the device mode to driver mode. Thanks!

**Comment by LinuxDonald** [comment]
With polychromatic git version it works really good. Thx for this pull request :)

**Comment by z3ntu** [comment]
Test results of this PR by someone:

> This branch of OpenRazer was confirmed to work on Ubuntu 20.04 using the
GUI of Polychromatic and RazerGenie. Lighting (color, effects, brightness),
DPI and polling rate could all be adjusted without issues. The only thing
that didn't work was opening the custom editor on RazerGenie, where a
"device type not implemented" error was generated. Furthermore, neither
PolyChromatic nor RazerGenie allowed adjusting the mouse button functions
or creating macros.

**Comment by mahmoudhossam** [comment]
As I haven't worked with the codebase before, I'm open to suggestions on how to improve this so it can be merged.

**Review by z3ntu — APPROVED** [review]
I still dislike the tilt wheel code but too much is being blocked by this PR so I'm accepting it nevertheless. Thanks for the PR!

(for reference: Commit before squashing: 618ff74b794a4b3b6d450a2293ffca06aaff58bd)

**Pull request merged**